### ISSUE TITLE
German Translation for MeshCentral

### ISSUE THREAD

**Issue by veitw** [Ylianst/MeshCentral#765]
Hi,

I thought I could spend some time over Christmas and New Year to contribute a German translation, but I am unable to get the Ressource Translation Tool up and running:

On Windows 10 Pro x64 1903 with .NET 3.5 and all updates installed, I see the following error when I click on a cell in the "Translated Text" column.
![image](https://user-images.githubusercontent.com/16285895/71516929-fbf6de00-28ab-11ea-9cad-4e98773b20f9.png)
The error messages translates to: The object of type "System.Collections.Generic.Dictionary`2[System.String,System.Object]" cannot be converted into type "System.String".

Also I cannot start the application on Linux using Mono/Xamarin, because WPF for .NET 3.0 is not implemented.

My third attempt was using Wine, but the win32 architecture target required to install either dotnet35sp1, dotnet35, dotnet30sp1 or dotnet30 winetricks targets seems to be b0rken in my 64bit Wine and all of them fail with one or another failure.

Is there a version of Ressource Translation Tool available that is built against a less WinXP-ish version of .NET and runs on either Windows 10 or Mono/Xamarin or Wine (e.g. dotnet40 installed ootb without problems)?

Best regards,
// Veit

**Comment by Ylianst** [comment]
Just posted a fixed [ResourceTranslator here](http://info.meshcentral.com/downloads/MeshCentral2/ResourceTranslator.zip). The problem was that the .json file I posted was not correct, it has an invalid entry at the top of the file in the first entry. I fixed the translate.json, but also fixed the translator tool to ignore any entries that are of invalid type.

Should work now. Let me know if it fixes it.

**Comment by veitw** [comment]
Seems to work, thank you. :)

**Comment by veitw** [comment]
Two questions:

1. Is there a short cut to jump to and select the next row that has no translation so far? This would speed-up the process quite much.

2. Where are singular/plural for entries such as " - Reset in {0} day{1}." configured? Because simply appending "s" for plural does not work for German. Or shall I remove the {1} placeholder and include the plural into the string, e.g. "{0} Tag(e)", "{0} Stunde(n)", "{0} Minute(n)"?

**Comment by veitw** [comment]
The tool seems to work fine now, but the JSON output is quite compact and thus not diffable against the original translation.json.
Which tool/commands do you recommend to bring it into the same format (and order) as the input file?

**Comment by Ylianst** [comment]
Hi.
1. There is no short cut to move to the next one, but I can work to add that.
2. I am working on fixing the singular/plural strings. Any string with that problem, let me know, I will fix it. it should be two English strings: " - Reset in {0} day." and " - Reset in {0} days."
3. Good point, I will try to fix the tool output to make it match.

**Comment by Ylianst** [comment]
FYI. I found all the strings with a replaced "s" and fixed them. It will be in the next published version: MeshCentral v0.4.6-t.

**Comment by veitw** [comment]
Hi @Ylianst,

as promised, here is the German translation for MeshCentral, based on the translate.json as of master on Monday, 30.12.2019: [translate.zip](https://github.com/Ylianst/MeshCentral/files/4016548/translate.zip)

Initially I planned to give you a pull request, but as the Translator's output does currently not diff nicely into its input and there have been changes to master since, thus causing a collision, I decided to send you the file directly for integration.

I clicked through my MeshCentral 0.4.6-s to verify the results and it looks quite good so far.
There are only minor problems such as the login screen box is a bit too small to fit the credential form. Also I used "Abbr." instead of "Abbrechen" for "Cancel", as the Cancel button seems currently to be fixed width.

Please leave me a note when the changes have been integrated, I will then download the new version and translate the strings you changed [here](#issuecomment-569771007).

Best regards,
// Veit

**Comment by Ylianst** [comment]
Thank you! That is amazing. I just published MeshCentral v0.4.6-v with your German translation. I put "Abbrechen" back for Cancel and did some adjustments. If you see any issues with text length, send over screen shots of the problems and I will work on getting that fixed.

I will work on getting the translation tool to output the correctly formatted JSON, I think that is a really good idea.

Also, I want to post your German translation on Twitter. Let me know how to credit you. Is it ok to say "Thank you to Veit for the MeshCentral German translation"?

Again, thank you so much on behalf of the community, it's great you did this.

**Comment by Ylianst** [comment]
Thank you! That is amazing. I just published MeshCentral v0.4.6-v with your German translation. I put "Abbrechen" back for Cancel and did some adjustments. If you see any issues with text length, send over screen shots of the problems and I will work on getting that fixed.

I will work on getting the translation tool to output the correctly formatted JSON, I think that is a really good idea.

Also, I want to post your German translation on Twitter. Let me know how to credit you. Is it ok to say "Thank you to Veit for the MeshCentral German translation"?

Again, thank you so much on behalf of the community, it's great you did this.

![MC2-German](https://user-images.githubusercontent.com/1319013/71701712-0586b700-2d80-11ea-81c2-a7bd871d0057.png)

**Comment by Ylianst** [comment]
I just updated the [Resource Translator](info.meshcentral.com/downloads/MeshCentral2/ResourceTranslator.zip) to include quick keys to go to previous/next empty entry (F9 and F10). I am not sure how I am going to make the output JSON look like the GitHub version... I still need to work on that.

**Comment by veitw** [comment]
Hi @Ylianst,

> Thank you! That is amazing. I just published MeshCentral v0.4.6-v with your German translation. I put "Abbrechen" back for Cancel and did some adjustments.

great, thanks. :)

> If you see any issues with text length, send over screen shots of the problems and I will work on getting that fixed.

This is the login form issue, that I believe most languages have, as their equivalent to "Username" is often longer:
![Login Screen](https://user-images.githubusercontent.com/16285895/71717509-9dd07a80-2e18-11ea-8cf6-3ca5a63cfdea.png)

> I will work on getting the translation tool to output the correctly formatted JSON, I think that is a really good idea.

Maybe instead offer instructions on how to bring it in shape? Using some multi platform pretty printer/sorter.

As an example I hacked this simple Perl 3-liner utilising the standard JSON.pm from CPAN:

```
!/usr/bin/env perl

use strict;
use warnings;
use JSON;

my $data = from_json(do{ local $/; <STDIN> }, {utf8 => 1});
$data->{strings} = [ sort{ $a->{en} cmp $b->{en} } @{$data->{strings}} ];
print to_json($data, {utf8 => 1, pretty => 1, space_before => 0, canonical => 1});
```

It reads JSON data from stdin and prints prettified and sorted to stdout.
There are caveats compared to your current standard format:
 - the sorting differs from yours
 - the indenting uses 3 spaces instead of 2 and JSON.pm does not seem to allow changing this (there are lots of other JSON modules, if this is a problem)
 - the language keys are sorted, so "en" is not always on top, but this might help diffing
 - Perl is not common on Windows systems, one needs to install e.g. Strawberry Perl

A file "translate-tidy.json" is included in the updated German translation ZIP archive below, so you may see how it looks after prettifying. The script is also included.

> Also, I want to post your German translation on Twitter. Let me know how to credit you. Is it ok to say "Thank you to Veit for the MeshCentral German translation"?

Of course you can. :) You may also use my full name as I do not have a Twitter handle to link to for reference.

> Again, thank you so much on behalf of the community, it's great you did this.

Always glad to help!

> ![MC2-German](https://user-images.githubusercontent.com/1319013/71701712-0586b700-2d80-11ea-81c2-a7bd871d0057.png)

Looks really nice, although I had to do some fine tuning when I saw the screenshots, e.g. capitalising adjective such as "Präsent" (Present) or "Eingeschaltet" (Powered) looks wrong to a German's eye as capitalising is reserved for nouns, names, polite form of personal pronouns and beginning of sentences.

> I just updated the [Resource Translator](info.meshcentral.com/downloads/MeshCentral2/ResourceTranslator.zip) to include quick keys to go to previous/next empty entry (F9 and F10). I am not sure how I am going to make the output JSON look like the GitHub version... I still need to work on that.

That helped a lot.

Beside doing some minor corrections, I just translated all the new strings added to master since 2019-12-30.

There are only very few strings are left that I need help with to have a full translation:

 - "Mesh Relay" -- is this a (new) software tool? I am not sure whether and how I shall translate this.
 - "PluginHandler could not event message: " -- there is a verb missing, thus I am not sure what it means.
 - "Relay teller" -- I think this is related to Mesh Relay, but I do not know how to translate "teller" meaningfully without context. It seams that nobody has translated this so far.

Also I intentionally did **not** translate the licenses from the terms into German, as I think even small mistakes or misinterpretions could lead to bigger problems here.
I know that some OSS licenses have official German translations that could be used here, but I do not think it is wise to use them here as licenses are subject to changes.
At least some of the license translations in the translation.json look translated automatically using Google Translate or similar and most likely contain translation errors, thus I strongly suggest to remove those translations, too.

[The updated translation for German](https://github.com/Ylianst/MeshCentral/files/4019038/translate2.zip) :)

Best regards,
// Veit

**Comment by petervanv** [comment]
@veitw "Relay teller"  was wrong adressed by me, i copyd the string but forgot to change in "nl"
Relay teller = relay count

**Comment by veitw** [comment]
> Relay teller = relay count

Ah yes, nl "teller" <-> de "Zähler", I could have guessed that. :D

**Comment by petervanv** [comment]
that tidy json looks awsome, i used all tree options to translate, but i prefer the clone git and than use pull request

**Comment by veitw** [comment]
Context question:

Does "duplicate" in these cases refer to the task of creating a duplicate of x, or to being a duplicate of x?
 - "Duplicate Agent"
 - "Duplicate User Group"
 - "Duplicate Group..."

Because each case translates differently in German, and I suspect in most other languages, too.

My guess from default.handlebars is:
 - "Duplicate Agent": being a duplicate
 - "Duplicate User Group": act of duplication
 - "Duplicate Group...": act of duplication

Maybe others fell for this, too...
I updated the translation ZIP four posts above to match these assumptions.

**Comment by Ylianst** [comment]
So much to do! I just updated the [ResourceTranslator](https://github.com/Ylianst/MeshCentral/issues/info.meshcentral.com/downloads/MeshCentral2/ResourceTranslator.zip) again to save JSON in the correct format. Now the problem is that all tools needs to save in the same format, space and same order. So, still more work to do.

Working on latest German translation now.

**Comment by Ylianst** [comment]
Just published MeshCentral v0.4.6-w with the latest German and Dutch translation fixes. I did not make any user interface fixes yet, but I need to get on that.

**Comment by Ylianst** [comment]
@veitw For posting on Twitter, if you want me to use your full name to give you credit... what is it? If you have a organization or university you want to promote, I can also include that.

**Comment by veitw** [comment]
> @veitw For posting on Twitter, if you want me to use your full name to give you credit... what is it? 

Oh, sorry, it's Veit Wahlich.

I think I used that only once, via E-Mail regarding the CIRA security issue. :)

**Comment by Ylianst** [comment]
@veitw Thanks. I just posted [my latest blog](https://meshcentral2.blogspot.com/2020/01/meshcentral-2019-review-and-new-features.html) and missed this, but you will be part of my next blog. Working on getting the [Twitter message](https://twitter.com/meshcentral) out now...

**Comment by Ylianst** [comment]
Just sent it out of Twitter: [https://twitter.com/MeshCentral/status/1213250144642031616](https://twitter.com/MeshCentral/status/1213250144642031616)

**Comment by veitw** [comment]
@Ylianst, thank you very much for appreciation. :)
